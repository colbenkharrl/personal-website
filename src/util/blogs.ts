import glob from 'fast-glob';
import { StaticImageData } from 'next/image';

interface Blog {
  title: string;
  description: string;
  author: string;
  date: string;
}

export interface BlogWithSlug extends Blog {
  slug: string;
  opengraphImage: StaticImageData;
}

async function importBlog(blogFilename: string): Promise<BlogWithSlug> {
  const { blog } = (await import(`../app/blog/${blogFilename}`)) as {
    default: React.ComponentType;
    blog: Blog;
  };

  const slug = blogFilename.replace(/(\/page)?\.mdx$/, '');

  const { default: opengraphImage } = (await import(
    `../app/blog/${slug}/opengraph-image.png`
  )) as {
    default: StaticImageData;
  };

  return {
    slug,
    opengraphImage,
    ...blog,
  };
}

export async function getAllBlogs() {
  const blogFilenames = await glob('*/page.mdx', {
    cwd: './src/app/blog',
  });

  const blogs = await Promise.all(blogFilenames.map(importBlog));

  return blogs.sort((a, z) => +new Date(z.date) - +new Date(a.date));
}
