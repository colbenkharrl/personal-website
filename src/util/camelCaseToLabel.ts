/**
 * Converts a camelCase string to a formatted label with spaces and each word capitalized.
 * @param camelCaseStr The camelCase string to be transformed.
 * @returns The formatted label.
 */
export const camelCaseToLabel = (camelCaseStr: string): string => {
  // Adjust regex to separate numbers and uppercase letters, treat sequences of numbers followed by lowercase as separate words
  const spacedStr = camelCaseStr
    .replace(/([0-9]+|[A-Z])/g, ' $1')
    .replace(/(\d+)([a-z])/g, '$1 $2');
  // Trim any leading space and capitalize each word
  return spacedStr.trim().replace(/\b\w/g, (char) => char.toUpperCase());
};
