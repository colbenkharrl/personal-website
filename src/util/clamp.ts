export function clamp(number: number, a: number, b: number) {
  let min = Math.min(a, b);
  let max = Math.max(a, b);
  return Math.min(Math.max(number, min), max);
}
