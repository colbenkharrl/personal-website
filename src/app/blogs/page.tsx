import { type Metadata } from 'next';

import { SimpleLayout } from '@/components/SimpleLayout';
import { getAllBlogs } from '@/util/blogs';
import { Blog } from '@/components/Blog';

export const metadata: Metadata = {
  title: 'Blog',
  description: 'My scattered thoughts about random things.',
};

export default async function BlogsIndex() {
  const blogs = await getAllBlogs();

  return (
    <SimpleLayout
      title="My scattered thoughts about random things."
      intro="I sometimes write down my ideas or document my experiences as I Roomba my way through my professional and personal life. A subset of those writings end up being published here for likely no one to enjoy. So, enjoy!"
    >
      <div className="md:border-l md:border-zinc-100 md:pl-6 md:dark:border-zinc-700/40">
        <div className="flex max-w-3xl flex-col space-y-16">
          {blogs.map((blog) => (
            <Blog key={blog.slug} blog={blog} />
          ))}
        </div>
      </div>
    </SimpleLayout>
  );
}
