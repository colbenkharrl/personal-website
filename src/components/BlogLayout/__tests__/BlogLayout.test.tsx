import { render, screen } from '@testing-library/react';
import { BlogLayout } from '../';
import { BlogWithSlug } from '@/util/blogs';

// mock useRouter
jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));

it('BlogLayout matches snapshot (regression test)', async () => {
  const testBlog: BlogWithSlug = {
    slug: 'test-blog',
    title: 'Test BlogLayout',
    description: 'Test blog description.',
    author: 'Matt Kharrl',
    date: '2023-11-09',
  };

  const blogContent = 'Test blog content';

  render(
    <BlogLayout blog={testBlog}>
      <p>{blogContent}</p>
    </BlogLayout>,
  );

  await screen.findByText(blogContent);

  expect(screen.getByTestId('Container')).toMatchSnapshot(
    'Full component snapshot.',
  );
});
