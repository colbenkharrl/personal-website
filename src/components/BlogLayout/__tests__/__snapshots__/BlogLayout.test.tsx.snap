// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogLayout matches snapshot (regression test): Full component snapshot. 1`] = `
<div
  class="sm:px-8 mt-16 lg:mt-32"
  data-testid="Container"
>
  <div
    class="mx-auto w-full max-w-7xl lg:px-8"
  >
    <div
      class="relative px-4 sm:px-8 lg:px-12"
    >
      <div
        class="mx-auto max-w-2xl lg:max-w-5xl"
      >
        <div
          class="xl:relative"
        >
          <div
            class="mx-auto max-w-2xl"
          >
            <article>
              <header
                class="flex flex-col"
              >
                <h1
                  class="mt-6 text-4xl font-bold tracking-tight text-zinc-800 dark:text-zinc-100 sm:text-5xl"
                >
                  Test BlogLayout
                </h1>
                <time
                  class="order-first flex items-center text-base text-zinc-400 dark:text-zinc-500"
                  datetime="2023-11-09"
                >
                  <span
                    class="h-4 w-0.5 rounded-full bg-zinc-200 dark:bg-zinc-500"
                  />
                  <span
                    class="ml-3"
                  >
                    November 9, 2023
                  </span>
                </time>
              </header>
              <div
                class="mt-8 prose dark:prose-invert"
                data-mdx-content="true"
              >
                <p>
                  Test blog content
                </p>
              </div>
            </article>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
